remove Plugins from Singleton get_state

jieru-hu opened the pull request:
<!-- Thank you for sending a PR and taking the time to improve Hydra -->

## Motivation

We do not need to pickle all of Singleton. change this to only pickle what is needed (`JobRuntime` and `OmegaConf` resolvers.)

This will also fix  that the plugin integration tests fails if other hydra tools/plugins is isntalled locally as well. 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes/No

## Test Plan
CI
locally
```
# install other plugin along with  ray launcher
FIX=1 PLUGINS=hydra_colorlog,hydra_ray_launcher  nox -s lint_plugins-3.8 test_plugins-3.8
```
## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

omry commented:
What I had in mind is that we:
1. fix get_state() to not return the the Plugins singleton state.
2. fix set_state() to call `Plugins.instance()` to (re)initialize the plugins.

omry commented on plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/_remote_invoke.py:
```diff
@@ -45,6 +46,7 @@ def launch_jobs(temp_dir: str) -> None:
                 )
             setup_globals()
             Singleton.set_state(singleton_state)
+            Plugins.instance()
```
And this inside set_state()?

omry commented on plugins/hydra_ray_launcher/hydra_plugins/hydra_ray_launcher/_core_aws.py:
```diff
@@ -90,6 +91,10 @@ def launch(
             sweep_configs.append(sweep_config)
 
+        state = Singleton.get_state()
+        del state.get("instances")[Plugins]
```
Why no do it inside get_state()?

jieru-hu replied:
see my comment below :)

jieru-hu commented:
> What I had in mind is that we:
> 
> 1. fix get_state() to not return the the Plugins singleton state.
> 2. fix set_state() to call `Plugins.instance()` to (re)initialize the plugins.

I see. This does seem better. I've fixed the PR to be along the same line of thoughts.

I'm not sure if you meant to actually change `Singleton` class. I've also tried that but ran into a circular imports issue. I didn't go further to make it work, since the issue I'm trying to fix here  is still on only `ray` plugin level.

more details below,
```
diff --git a/hydra/core/singleton.py b/hydra/core/singleton.py
index d184a4e29..13bd5a9db 100644
--- a/hydra/core/singleton.py
+++ b/hydra/core/singleton.py
@@ -4,6 +4,8 @@ from typing import Any, Dict
 
 from omegaconf.basecontainer import BaseContainer
 
+from hydra.core.plugins import Plugins
+
 
 class Singleton(type):
     _instances: Dict[type, "Singleton"] = {}
@@ -18,6 +20,7 @@ class Singleton(type):
 
     @staticmethod
     def get_state() -> Any:
+        del Singleton.get_state().get("instances")[Plugins]
         return {
             "instances": Singleton._instances,
             "omegaconf_resolvers": deepcopy(BaseContainer._resolvers),
@@ -26,4 +29,5 @@ class Singleton(type):
     @staticmethod
     def set_state(state: Any) -> None:
         Singleton._instances = state["instances"]
+        Plugins.instance()
         BaseContainer._resolvers = deepcopy(state["omegaconf_resolvers"])



Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/__init__.py", line 5, in <module>
    from hydra import utils
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/utils.py", line 12, in <module>
    from hydra._internal.utils import (
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/utils.py", line 19, in <module>
    from hydra.core.utils import get_valid_filename, split_config_path
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/core/utils.py", line 17, in <module>
    from hydra.core.hydra_config import HydraConfig
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/core/hydra_config.py", line 6, in <module>
    from hydra.conf import HydraConf
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/conf/__init__.py", line 7, in <module>
    from hydra.core.config_store import ConfigStore
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/core/config_store.py", line 9, in <module>
    from hydra.core.singleton import Singleton
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/core/singleton.py", line 7, in <module>
    from hydra.core.plugins import Plugins
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/core/plugins.py", line 14, in <module>
    from hydra._internal.sources_registry import SourcesRegistry
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/sources_registry.py", line 4, in <module>
    from hydra.core.singleton import Singleton
ImportError: cannot import name 'Singleton' from partially initialized module 'hydra.core.singleton' (most likely due to a circular import) (/Users/jieru/workspace/hydra-fork/hydra/hydra/core/singleton.py)

```

jieru-hu commented:
per offline discussion  - move imports to local :)

omry commented on hydra/core/singleton.py:
```diff
@@ -18,6 +18,12 @@ def instance(cls: Any, *args: Any, **kwargs: Any) -> Any:
 
     @staticmethod
     def get_state() -> Any:
+        try:
+            from hydra.core.plugins import Plugins
+
+            del Singleton._instances[Plugins]
+        except KeyError:
+            pass
```
You are changing the actual singleton state, don't do that.
Make a copy of the map without the Plugins singleton.

The only reason this did not cause issues is that plugin interaction is currently only happening on startup, but this can change.

jieru-hu commented:
>You are changing the actual singleton state, don't do that.
Make a copy of the map without the Plugins singleton.
The only reason this did not cause issues is that plugin interaction is currently only happening on startup, but this can change.


I see. Thanks! updated the PR per your comment, also added a unit test.

jieru-hu commented:
rebase...

omry commented on hydra/core/singleton.py:
```diff
@@ -18,12 +18,22 @@ def instance(cls: Any, *args: Any, **kwargs: Any) -> Any:
     @staticmethod
     def get_state() -> Any:
+        instances = deepcopy(Singleton._instances)
+        try:
+            from hydra.core.plugins import Plugins
+
+            del instances[Plugins]
+        except KeyError:
+            pass
```
try:
```python
 instances.pop(Plugins, None)
```

jieru-hu replied:
Nice!!

omry commented on hydra/core/singleton.py:
```diff
@@ -18,12 +18,22 @@ def instance(cls: Any, *args: Any, **kwargs: Any) -> Any:
         Singleton._instances = state["instances"]
+        from hydra.core.plugins import Plugins
+
+        Plugins.instance()
```
Add a comment explaining what this is doing.

omry commented on hydra/core/singleton.py:
```diff
@@ -18,12 +18,21 @@ def instance(cls: Any, *args: Any, **kwargs: Any) -> Any:
 
     @staticmethod
     def get_state() -> Any:
+        instances = deepcopy(Singleton._instances)
+        # Plugins can cause issues for pickling Singleton.get_state().
+        from hydra.core.plugins import Plugins
+
+        instances.pop(Plugins, None)
```
```suggestion
        instances = deepcopy(Singleton._instances)
```suggestion
        # Plugins can cause issues for pickling the singleton state
        # Exclude them and re-initialize them on set_state()
        from hydra.core.plugins import Plugins

        instances.pop(Plugins, None)
```

omry commented on plugins/hydra_ray_launcher/tests/test_ray_aws_launcher.py:
```diff
@@ -108,22 +108,7 @@
-
-def build_plugin_wheel(tmp_wheel_dir: str) -> str:
+def build_ray_launcher_wheel(tmp_wheel_dir: str) -> str:
     chdir_hydra_root()
     plugin = "hydra_ray_launcher"
     os.chdir(Path("plugins") / plugin)
```
You can probably use monkeypatch to change the directory and restore automatically.

jieru-hu replied:
I didn't know `monkeypatch` could do that, I thought we usually do `fixture`?

omry replied:
https://github.com/facebookresearch/hydra/blob/master/tests/test_hydra.py#L1049-L1050

jieru-hu replied:
I see, thanks!

omry commented on hydra/core/singleton.py:
```diff
@@ -18,12 +18,21 @@ def instance(cls: Any, *args: Any, **kwargs: Any) -> Any:
     @staticmethod
     def set_state(state: Any) -> None:
         Singleton._instances = state["instances"]
+        # Re-init Plugins since it was removed  in get_state()
```
```suggestion
        # Reinitialize the the Plugin singleton (discover all plugins etc).
```

omry reviewed (approved):
Merge once CI passes.
